#165376778 Add fields to user and asset models.

RonKbS opened the pull request:
## What does this PR do?
This PR adds department fields to user and asset tables. It also adds active and paid_postpaid fields to the asset table as well. A table is also added to track owners of assets, different from people to whom assets are assigned. Third, the PR restricts users to only view assets in their departments and location

## Description of Task to be completed?
Currently, tracking of which departments users and assets belong to is not being done. There is also no tracking of actual owners (department, user or workspace) of an asset but only an assignee. Thirdly, users can view assets irrespective of which departments or locations they belong. The changes in this PR address these issues.

## How should this be manually tested?
1. On the browser, go to URL `http://127.0.0.1:8000/api/v1/users/` and add a user with an additional key-value for `department` for an existent department. A response with the given department should be returned

2. Go to `http://127.0.0.1:8000/api/v1/assets/` and post an asset with additional key-value pairs for `department`, `active` and `prepaid_postpaid`. A response with the same data should be returned.

3. If there are any existing assets not in the same department or location as a user interacting with `http://127.0.0.1:8000/api/v1/assets/` using a get method, these assets will not appear in the response.

4. Go to `http://127.0.0.1:8000/api/v1/manage-assets/{uuid}/` and patch an asset's active, prepaid_postpaid or department fields. A response with the updated asset field should be returned. This is dependent on which type of asset is being worked on, with only simcards and mifis allowed to have their active and prepaid_postpaid fields.

5. Endpoints that were previously using current_owner/previous_owner are now using current_assignee/previous_assignee.

## What are the relevant pivotal tracker stories?
[#165376778](https://www.pivotaltracker.com/story/show/165376778)

RonKbS committed:
feat(department): add fields to user and asset models

- add foreign key fields to users and assets to department table
- add table for asset owners and link it to asset table
- add active_inactive and paid_postpaid fields to asset table
- restrict users to view assets in their locations and departments only

[Finishes #165376778]

egichuri commented on api/views/assets.py:
```diff
@@ -111,6 +111,10 @@ class AssetViewSet(ModelViewSet):
+
+        # view only assets in ones location and ones department
+        query_filter = {"department": user.department, "asset_location": user.location}
+
```
this should be on `ManageAssetViewSet`

egichuri commented on core/constants.py:
```diff
@@ -111,3 +111,8 @@
     (DAMAGED, "Damaged"),
 )
+
+PAID = "paid"
```
prepaid

egichuri commented on api/views/assets.py:
```diff
@@ -111,6 +111,10 @@ class AssetViewSet(ModelViewSet):
         query_filter = {}
+
+        # view only assets in ones location and ones department
+        query_filter = {"department": user.department, "asset_location": user.location}
```
this should be on `ManageAssetViewSet`

egichuri replied:
for clarity - `ManageAssetViewSet` is for admin users - the one we use for the FE dashboard, `AssetViewSet` is for security users (for checkin/checkout) and andela users (to access assets assigned to them.)

RonKbS replied:
The PT story does not specify which user_type is to be restricted to seeing assets for their teams and locations. So to confirm, admin users will be the ones with this restriction?

egichuri replied:
yeah.. only admin users.

RonKbS committed:
feat(department): add fields to user and asset models

- adjust current_owner and previous_owner fields to current_assignee and
    previous_assignee

[Finishes #165376778]

RonKbS committed:
feat(department): add fields to user and asset models

- move department restriction to manage_asset_viewset
- change paid_postpaid field to prepaid_postpaid

[Finishes #165376778]

RonKbS committed:
feat(department): add fields to user and asset models

- add restriction for which types of assets can have active and
    prepaid_postpaid fields edited

[Finishes #165376778]

RonKbS committed:
- add expiry_date field to asset_serializer

[Finishes #165376778]

kmazi commented:
@egichuri How about this one, should I go on to merge?

MadgeKinyanjui commented on api/serializers/assets.py:
```diff
@@ -128,6 +140,43 @@ def to_internal_value(self, data):
+                "prepaid_or_postpaid": "Only sim cards can be prepaid or postpaid"
+            },
+            "mifi": {"active": "Only mifi cards can be activated or deactivated"},
+            "atmcard": {"expiry_date": "Only atm cards can have this field updated"},
```
@RonKbS change this to emburse card and should add the field for active or inactive

MadgeKinyanjui commented on api/serializers/assets.py:
```diff
@@ -128,6 +140,43 @@ def to_internal_value(self, data):
+        expiry_date = validated_data.get('expiry_date')
+        instance_type_and_associated_error = {
+            "simcard": {
+                "prepaid_or_postpaid": "Only sim cards can be prepaid or postpaid"
```
Sim cards in place of active or inactive have a filed called 'plan' that has three options basic plan, business plan and none.

RonKbS committed:
- reformat migration file to follow recently added migrations

RonKbS committed:
- change prepaid_postpaid field to paid field with associated options
- change atmcard to embusecard

RonKbS commented:
@MadgeKinyanjui, the asked for changes have been implemented.
